# Hands-on example

In [1]:
import os
import warnings
import sys

import pandas as pd
import numpy as np

from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from sklearn.model_selection import train_test_split
from sklearn.linear_model import ElasticNet

import mlflow
import mlflow.sklearn

## 0. The data

* The data set used in this example is from http://archive.ics.uci.edu/ml/datasets/Wine+Quality
* P. Cortez, A. Cerdeira, F. Almeida, T. Matos and J. Reis.
* Modeling wine preferences by data mining from physicochemical properties. In Decision Support Systems, Elsevier, 47(4):547-553, 2009.

In [2]:
data_path = "data/wine-quality.csv"
data = pd.read_csv(data_path)

data.sample(10)


fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
4522            9.0              0.20         0.33             3.5      0.049   
3867            7.3              0.25         0.26             7.2      0.048   
2517            7.3              0.51         0.29            11.3      0.034   
2760            7.7              0.16         0.36             5.9      0.054   
400             6.3              0.20         0.40             1.5      0.037   
4629            5.6              0.24         0.34             2.0      0.041   
3028            7.5              0.18         0.37             6.2      0.050   
1891            7.6              0.33         0.41            13.7      0.045   
3848            7.1              0.39         0.79             1.4      0.194   
4406            6.3              0.29         0.29             3.3      0.037   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
4522                 10.0                  40.0  0.99440  3.14       0.36   
3867                 52.0                 207.0  0.99587  3.12       0.37   
2517                 61.0                 224.0  0.99683  3.14       0.56   
2760                 25.0                 148.0  0.99578  3.25       0.54   
400                  35.0                 107.0  0.99170  3.46       0.50   
4629                 14.0                  73.0  0.98981  3.04       0.45   
3028                 21.0                 138.0  0.99546  3.20       0.55   
1891                 44.0                 197.0  0.99890  3.18       0.64   
3848                 23.0                  90.0  0.99212  3.17       0.46   
4406                 32.0                 140.0  0.98950  3.17       0.36   

      alcohol  quality  
4522      9.8        6  
3867      9.2        5  
2517      9.5        6  
2760     10.2        6  
400      11.4        6  
4629     11.6        7  
3028     10.5        6  
1891      9.1        5  
3848     10.5        6  
4406     12.8        7

## 1. Tracking experiments

### Tracking stores
MLflow supports two types of backend stores: *file store* and *database-backed* store.

- Local file path (specified as file:/my/local/dir), where data is just directly stored locally. Defaults to `mlruns/`
- Database encoded as <dialect>+<driver>://<username>:<password>@<host>:<port>/<database>. Mlflow supports the dialects mysql, mssql, sqlite, and postgresql. For more details, see SQLAlchemy database uri.
- HTTP server (specified as https://my-server:5000), which is a server hosting an MLFlow tracking server.
- Databricks workspace (specified as databricks or as databricks://<profileName>, a Databricks CLI profile.
    
### Artifact stores
- Amazon S3
- Azure Blob Storage
- Google Cloud Storage
- FTP server
- SFTP Server
- NFS
- HDFS

Start the MLflow tracking server by 

```
mlflow server \
    --backend-store-uri /mnt/persistent-disk \
    --default-artifact-root s3://my-mlflow-bucket/ \
    --host 0.0.0.0
    --port 5000
```

or use the default storage method to write to `mlruns/`.

In [3]:
# mlflow server --backend-store-uri mlruns/ --default-artifact-root mlruns/ --host 0.0.0.0 --port 5000
remote_server_uri = "http://0.0.0.0:5000" # set to your server URI
mlflow.set_tracking_uri(remote_server_uri)  # or set the MLFLOW_TRACKING_URI in the env

In [4]:
mlflow.tracking.get_tracking_uri()

'http://0.0.0.0:5000'

In [5]:
exp_name = "ElasticNet_wine"
mlflow.set_experiment(exp_name)

### What do we track?

- **Code Version**: Git commit hash used for the run (if it was run from an MLflow Project)
- **Start & End Time**: Start and end time of the run
- **Source**: what code run?
- **Parameters**: Key-value input parameters.
- **Metrics**: Key-value metrics, where the value is numeric (can be updated over the run)
- **Artifacts**: Output files in any format.

In [6]:
def eval_metrics(actual, pred):
    # compute relevant metrics
    rmse = np.sqrt(mean_squared_error(actual, pred))
    mae = mean_absolute_error(actual, pred)
    r2 = r2_score(actual, pred)
    return rmse, mae, r2


def load_data(data_path):
    data = pd.read_csv(data_path)

    # Split the data into training and test sets. (0.75, 0.25) split.
    train, test = train_test_split(data)

    # The predicted column is "quality" which is a scalar from [3, 9]
    train_x = train.drop(["quality"], axis=1)
    test_x = test.drop(["quality"], axis=1)
    train_y = train[["quality"]]
    test_y = test[["quality"]]
    return train_x, train_y, test_x, test_y

def log_diff():
    with open("diff", "w") as f:
        diff = os.popen("git diff")
        f.write(diff.read())
    mlflow.log_artifact("diff")


def train(alpha=0.5, l1_ratio=0.5):
   
    # train a model with given parameters
    warnings.filterwarnings("ignore")
    np.random.seed(40)

    # Read the wine-quality csv file (make sure you're running this from the root of MLflow!)
    data_path = "data/wine-quality.csv"
    train_x, train_y, test_x, test_y = load_data(data_path)

    # Useful for multiple runs (only doing one run in this sample notebook)    
    with mlflow.start_run():
        # Execute ElasticNet
        lr = ElasticNet(alpha=alpha, l1_ratio=l1_ratio, random_state=42)
        lr.fit(train_x, train_y)

        # Evaluate Metrics
        predicted_qualities = lr.predict(test_x)
        (rmse, mae, r2) = eval_metrics(test_y, predicted_qualities)

        # Print out metrics
        print("Elasticnet model (alpha=%f, l1_ratio=%f):" % (alpha, l1_ratio))
        print("  RMSE: %s" % rmse)
        print("  MAE: %s" % mae)
        print("  R2: %s" % r2)

        # Log parameter, metrics, and model to MLflow
        mlflow.log_param(key="alpha", value=alpha)
        mlflow.log_param(key="l1_ratio", value=l1_ratio)
        mlflow.log_metric(key="rmse", value=rmse)
        mlflow.log_metrics({"mae": mae, "r2": r2})
        mlflow.log_artifact(data_path)
        print("Save to: {}".format(mlflow.get_artifact_uri()))
        
        mlflow.sklearn.log_model(lr, "model")

        sha = os.popen("git rev-parse --short HEAD")
        branch = os.popen("git rev-parse --abbrev-ref HEAD")
        mlflow.set_tag('mlflow.source.git.commit', sha.read())
        mlflow.set_tag('mlflow.source.git.branch', branch.read())
        log_diff()
        
        mlflow.end_run()


In [ ]:
# could also do
# with mlflow.start_run():
#     for epoch in range(0, 3):
#          mlflow.log_metric(key="quality", value=2*epoch, step=epoch)


In [7]:
train(0.5, 0.5)


Elasticnet model (alpha=0.500000, l1_ratio=0.500000):
  RMSE: 0.82224284975954
  MAE: 0.6278761410160693
  R2: 0.12678721972772689
Save to: file:///cargo/dev/AMLD_2020_MLflow/mlruns/1/9e58ec9cb01c4dcbb8366677bb32825e/artifacts


In [8]:
train(0.2, 0.2)

Elasticnet model (alpha=0.200000, l1_ratio=0.200000):
  RMSE: 0.7859129997062342
  MAE: 0.6155290394093895
  R2: 0.2022463182289208
Save to: file:///cargo/dev/AMLD_2020_MLflow/mlruns/1/fd5ec88cd2e2431d9b3ee9c5c7fc1aa6/artifacts


In [9]:
train(0.1, 0.1)

Elasticnet model (alpha=0.100000, l1_ratio=0.100000):
  RMSE: 0.7792546522251949
  MAE: 0.6112547988118586
  R2: 0.2157063843066196
Save to: file:///cargo/dev/AMLD_2020_MLflow/mlruns/1/5fcf9cd91e624169a69a80fd680bd743/artifacts


### 1.1 Comparing runs
Run `mlflow ui` in a terminal or `http://your-tracking-server-host:5000` to view the experiment log and visualize and compare different runs and experiments. The logs and the model artifacts are saved in the `mlruns` directory (or where you specified).

## 2. Packaging the experiment as a MLflow project as conda env

Specify the entrypoint for this project by creating a `MLproject` file and adding an conda environment with a `conda.yaml`. You can copy the yaml file from the experiment logs.

To run this project, invoke `mlflow run . -P alpha=0.42`. After running this command, MLflow runs your training code in a new Conda environment with the dependencies specified in `conda.yaml`.


## 3. Deploy the model

Deploy the model locally by running 

`mlflow models serve -m mlruns/0/f5f7c052ddc5469a852aa52c14cabdf1/artifacts/model/ -h 0.0.0.0 -p 1234`

Test the endpoint:

`curl -X POST -H "Content-Type:application/json; format=pandas-split" --data '{"columns":["alcohol", "chlorides", "citric acid", "density", "fixed acidity", "free sulfur dioxide", "pH", "residual sugar", "sulphates", "total sulfur dioxide", "volatile acidity"],"data":[[12.8, 0.029, 0.48, 0.98, 6.2, 29, 3.33, 1.2, 0.39, 75, 0.66]]}' http://0.0.0.0:1234/invocations`

You can also simply build a docker image from your model

`mlflow models build-docker -m mlruns/1/d671f37a9c7f478989e67eb4ff4d1dac/artifacts/model/ -n elastic_net_wine`

and run the container with

`docker run -p 8080:8080 elastic_net_wine`.

Or you can directly deploy to AWS sagemaker or Microsoft Azure ML.

## 4. Tagging runs

In [10]:
from datetime import datetime
from mlflow.tracking import MlflowClient

client = MlflowClient()
experiments = client.list_experiments() # returns a list of mlflow.entities.Experiment
print(experiments)

[<Experiment: artifact_location='file:///cargo/dev/AMLD_2020_MLflow/mlruns/1', experiment_id='1', lifecycle_stage='active', name='ElasticNet_wine', tags={}>, <Experiment: artifact_location='file:///cargo/dev/AMLD_2020_MLflow/mlruns/0', experiment_id='0', lifecycle_stage='active', name='Default', tags={}>]


In [12]:
# get the run
_run = client.get_run(run_id="5fcf9cd91e624169a69a80fd680bd743")
print(_run)

<Run: data=<RunData: metrics={'mae': 0.6112547988118586,
 'r2': 0.2157063843066196,
 'rmse': 0.7792546522251949}, params={'alpha': '0.1', 'l1_ratio': '0.1'}, tags={'mlflow.log-model.history': '[{"run_id": "5fcf9cd91e624169a69a80fd680bd743", '
                             '"artifact_path": "model", "utc_time_created": '
                             '"2020-03-22 22:58:49.392092", "flavors": '
                             '{"python_function": {"loader_module": '
                             '"mlflow.sklearn", "python_version": "3.6.9", '
                             '"data": "model.pkl", "env": "conda.yaml"}, '
                             '"sklearn": {"pickled_model": "model.pkl", '
                             '"sklearn_version": "0.22.2.post1", '
                             '"serialization_format": "cloudpickle"}}}]',
 'mlflow.source.git.branch': 'master\n',
 'mlflow.source.git.commit': '5e8fe91\n',
 'mlflow.source.name': '/cargo/home/.venvs/mlflow-pydata/lib/python3.6/site-packages/i

In [ ]:
# add a tag to the run
dt = datetime.now().strftime("%d-%m-%Y (%H:%M:%S.%f)")
client.set_tag(_run.info.run_id, "deployed", dt)

# Connect to a postgesql db

Connect to the db:
    
`sudo -u postgres psql`

Create a user and a database for the tracking server:
    
```
CREATE DATABASE mlflow;
CREATE USER mlflow WITH ENCRYPTED PASSWORD 'mlflow';
GRANT ALL PRIVILEGES ON DATABASE mlflow TO mlflow;
```

```
mlflow server --backend-store-uri postgresql://mlflow:mlflow@localhost/mlflow --default-artifact-root file:/home/tobias/Projects/mlflow_talk/pydataberlin-2019/mlruns/ -h 0.0.0.0 -p 8000
```

Run the notebook again with this tracking server.

Look at the db:

`psql mlflow`

`SELECT * FROM experiments;`

`SELECT * FROM runs;`